pkg/solana/logpoller/log_poller.go

```diff
@@ -0,0 +1,316 @@
+	}
+
+	delete(lp.filtersByName, filter.Name)
+	filtersByEventSig, ok := lp.filtersByAddress[filter.Address]
+	if !ok {
+		lp.lggr.Warnw("Filter not found in filtersByAddress", "name", name, "address", filter.Address)
+		return nil
+	}
+
+	filtersByEventSig[filter.EventSig], ok = removeFilterFromSlice(filtersByEventSig[filter.EventSig], filter)
+	if !ok {
+		lp.lggr.Warnw("Filter not found in filtersByEventSig", "name", name, "address", filter.Address)
+	}
+
+	if len(filtersByEventSig[filter.EventSig]) == 0 {
+		delete(filtersByEventSig, filter.EventSig)
+	}
+
+	if len(lp.filtersByAddress[filter.Address]) == 0 {
+		delete(lp.filtersByAddress, filter.Address)
+	}
+
+	// remove or ensure that filters was not present in the slice to backfill
+	lp.filtersToBackfill, _ = removeFilterFromSlice(lp.filtersToBackfill, filter)
+	lp.filtersToDelete = append(lp.filtersToDelete, filter)
+	return nil
+}
+
+func (lp *LogPoller) loadFilters(ctx context.Context) error {
+	if lp.loadedFilters.Load() {
+		return nil
+	}
+
+	lp.lggr.Debugw("Loading filters from db")
+	lp.filtersMutex.Lock()
+	defer lp.filtersMutex.Unlock()
+	// reset filters' indexes to ensure we do not have partial data from the previous run
+	lp.filtersByAddress = make(map[PublicKey]map[EventSignature][]Filter)
+	lp.filtersByName = make(map[string]Filter)
+	lp.filtersToBackfill = nil
+	lp.filtersToDelete = nil
+
+	ctx, cancel := lp.chStop.Ctx(ctx)
+	defer cancel()
+	filters, err := lp.orm.SelectFilters(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to select filters from db: %w", err)
+	}
+
+	for _, filter := range filters {
+		if filter.IsDeleted {
+			lp.filtersToDelete = append(lp.filtersToDelete, filter)
+			continue
+		}
+
+		if _, ok := lp.filtersByName[filter.Name]; ok {
+			errMsg := fmt.Sprintf("invariant violation while loading from db: expected filters to have unique name: %s ", filter.Name)
+			lp.lggr.Critical(errMsg)
+			return errors.New(errMsg)
+		}
+
+		lp.filtersByName[filter.Name] = filter
+		filtersByEventSig, ok := lp.filtersByAddress[filter.Address]
+		if !ok {
+			filtersByEventSig = make(map[EventSignature][]Filter)
+			lp.filtersByAddress[filter.Address] = filtersByEventSig
+		}
+
+		filtersByEventSig[filter.EventSig] = append(filtersByEventSig[filter.EventSig], filter)
+		lp.filtersToBackfill = append(lp.filtersToBackfill, filter)
+	}
+
+	lp.loadedFilters.Store(true)
+	return nil
+}
+
+func removeFilterFromSlice(filters []Filter, filter Filter) ([]Filter, bool) {
+	index := slices.IndexFunc(filters, func(item Filter) bool {
+		return item.ID == filter.ID
+	})
+	if index == -1 {
+		return filters, false
+	}
+
+	lastIdx := len(filters) - 1
+	filters[index], filters[lastIdx] = filters[lastIdx], filters[index]
+	return filters[:lastIdx], true
+}
+
+// matchingFilters - allows to iterate through filters that match provided keys
+func (lp *LogPoller) matchingFilters(addr PublicKey, eventSignature EventSignature) iter.Seq[Filter] {
```

Comment:
I don't see this used anywhere in the code, I guess it's intended for future use? Should we have a test for it?